### PR TITLE
Cursor position should be restored after change (with React fix)

### PR DESCRIPTION
**Pull request by mdziekon**
Hello everyone, since there's been no activity in [this PR](https://github.com/nosir/cleave.js/pull/150) created by @reborg (PR related to this issue: #25), I've fixed the problem with React build on my own and pushed it from my own fork. I'm not sure if this is a proper way of "fixing" other person's PR, so if this PR is violating some contribution rules, please contact me and let me know how should I do it properly.

**About the fix itself**:

I've modified how React component changes the selection modification.  
First of all - React component do not have DOM elements bindings by default, so I had to modify the ``htmlRef`` part (making it store the ref locally and then apply callback from the outside world, provided by the lib user).  
The second part of the fix was to properly set the selection - since React's ``setState`` is not a synchronous call, one cannot modify the input's state inside the same method that changes the component's state (as it will happen **before** the state change). Of course, this could be deferred using ``setTimeout``, but I've done this by using the lifecycle method React provides (``componentDidUpdate`` in this case), which apparently works slightly better ("temporal cursor's jump" were almost non-existent).

### CODE REVIEW

**Comment by Adrian0350** [comment]
This should really be tested and merged.

IT IS an important update.

**Comment by andela-omesanni** [comment]
it would really help if this PR gets merged. I need this fix in a project am currently working on

File: src/Cleave.react.js [hunk 1]
```diff
@@ -308,7 +361,15 @@ var Cleave = React.createClass({
...
                 type="text"
-                ref={htmlRef}
+                ref={(ref) => {
+                    this.element = ref;
+
+                    if (!htmlRef) {
+                        return;
+                    }
+
+                    htmlRef.apply(this, arguments);
```
Comment (silasb) [inline src/Cleave.react.js]: I'm pretty sure this won't work.  We don't want to change the `this` of `htmlRef`.

Comment (mdziekon) [reply src/Cleave.react.js]: The last line ``htmlRef.apply(this, arguments);`` is roughly the same as we would pass ``htmlRef`` directly as it was done in the previous version, so this line should be fine.

However, the problem is with the arrow function that I've mistakenly used (which in this case binds a different value to ``this``, namely the component itself). The arrow function should be replaced with a regular anonymous function an upper scope's ``this`` capture, to use with ``this.element``.

I'm not sure if this is what you've suggested with your comment, but nevertheless thanks for pointing out there might be a problem here.

**Comment by nosir** [comment]
Good work @mdziekon Thanks for the contribution.
Merged.

**Pull request merged**